File: src/schemas/jobs.py
"""Job posting schemas."""

from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field, field_validator, model_validator

from src.enums import JobStatus

JOB_SHORT_DESC_MAX = 140
JOB_TAG_MAX_LEN = 30
JOB_TAG_MAX_COUNT = 6
JOB_REQ_TEXT_MAX = 200
JOB_REQ_MIN_COUNT = 3
JOB_REQ_MAX_COUNT = 15


class JobRequirementItem(BaseModel):
    """A single requirement bullet."""

    text: str = Field(..., min_length=1, max_length=JOB_REQ_TEXT_MAX)

    @field_validator("text")
    @classmethod
    def _strip(cls, v: str) -> str:
        stripped = v.strip()
        if not stripped:
            raise ValueError("requirement text cannot be empty")
        return stripped


def _validate_requirements(
    v: list[JobRequirementItem],
) -> list[JobRequirementItem]:
    if not (JOB_REQ_MIN_COUNT <= len(v) <= JOB_REQ_MAX_COUNT):
        raise ValueError(
            f"requirements must have between {JOB_REQ_MIN_COUNT} and "
            f"{JOB_REQ_MAX_COUNT} items"
        )
    return v


def _validate_tags(v: list[str]) -> list[str]:
    cleaned: list[str] = []
    seen: set[str] = set()
    for raw in v:
        tag = raw.strip()
        if not tag:
            continue
        if len(tag) > JOB_TAG_MAX_LEN:
            raise ValueError(f"tag must be at most {JOB_TAG_MAX_LEN} characters")
        key = tag.lower()
        if key in seen:
            continue
        seen.add(key)
        cleaned.append(tag)
    if len(cleaned) > JOB_TAG_MAX_COUNT:
        raise ValueError(f"at most {JOB_TAG_MAX_COUNT} tags allowed")
    return cleaned


class JobCreate(BaseModel):
    """Schema for creating a job posting."""

    title: str = Field(..., max_length=100)
    short_description: str = Field(..., min_length=1, max_length=JOB_SHORT_DESC_MAX)
    description: str = Field(..., max_length=5000)
    requirements: list[JobRequirementItem]
    tags: list[str] = Field(default_factory=list)
    location: str = Field(..., max_length=100)
    salary_min: int = Field(..., ge=0)
    salary_max: int = Field(..., ge=0)

    _validate_requirements = field_validator("requirements")(_validate_requirements)
    _validate_tags = field_validator("tags")(_validate_tags)

    @model_validator(mode="after")
    def _check_salary_range(self):
        if self.salary_min > self.salary_max:
            raise ValueError("salary_min must be <= salary_max")
        return self


class JobUpdate(BaseModel):
    """Schema for a company updating their job posting.

    Fields backed by NOT NULL columns reject an explicit ``null`` so a PATCH
    cannot bypass the database constraint. ``is_featured`` is intentionally
    omitted — only admins toggle that flag (see ``JobAdminUpdate``).
    """

    title: str | None = Field(None, max_length=100)
    short_description: str | None = Field(
        None, min_length=1, max_length=JOB_SHORT_DESC_MAX
    )
    description: str | None = Field(None, max_length=5000)
    requirements: list[JobRequirementItem] | None = None
    tags: list[str] | None = None
    location: str | None = Field(None, max_length=100)
    salary_min: int | None = Field(None, ge=0)
    salary_max: int | None = Field(None, ge=0)
    status: JobStatus | None = None

    @field_validator(
        "title",
        "short_description",
        "description",
        "requirements",
        "tags",
        "location",
        "salary_min",
        "salary_max",
    )
    @classmethod
    def reject_explicit_null(cls, v: object) -> object:
        if v is None:
            raise ValueError("Field cannot be set to null on update")
        return v

    @field_validator("requirements")
    @classmethod
    def _validate_requirements_list(
        cls, v: list[JobRequirementItem]
    ) -> list[JobRequirementItem]:
        return _validate_requirements(v)

    @field_validator("tags")
    @classmethod
    def _validate_tags_list(cls, v: list[str]) -> list[str]:
        return _validate_tags(v)

    @model_validator(mode="after")
    def _check_salary_range(self):
        # Only enforce when BOTH bounds are being set in the same request;
        # partial updates that change only one bound rely on the DB CHECK
        # constraint ck_job_salary_range as the safety net.
        if (
            self.salary_min is not None
            and self.salary_max is not None
            and self.salary_min > self.salary_max
        ):
            raise ValueError("salary_min must be <= salary_max")
        return self


class JobAdminCreate(BaseModel):
    """Schema for an admin creating a job posting against a specific company."""

    company_id: int
    title: str = Field(..., max_length=100)
    short_description: str = Field(..., min_length=1, max_length=JOB_SHORT_DESC_MAX)
    description: str = Field(..., max_length=5000)
    requirements: list[JobRequirementItem]
    tags: list[str] = Field(default_factory=list)
    is_featured: bool = False
    location: str = Field(..., max_length=100)
    salary_min: int = Field(..., ge=0)
    salary_max: int = Field(..., ge=0)
    status: JobStatus = JobStatus.PUBLISHED

    _validate_requirements = field_validator("requirements")(_validate_requirements)
    _validate_tags = field_validator("tags")(_validate_tags)

    @model_validator(mode="after")
    def _check_salary_range(self):
        if self.salary_min > self.salary_max:
            raise ValueError("salary_min must be <= salary_max")
        return self


class JobAdminUpdate(JobUpdate):
    """Admin-only update schema. Adds ``is_featured`` toggle."""

    is_featured: bool | None = None

    @field_validator("is_featured")
    @classmethod
    def _reject_null_featured(cls, v: bool | None) -> bool:
        if v is None:
            raise ValueError("Field cannot be set to null on update")
        return v


class JobRead(BaseModel):
    """Schema for reading job data."""

    model_config = ConfigDict(from_attributes=True)

    id: int
    company_id: int
    title: str
    short_description: str
    description: str
    requirements: list[JobRequirementItem]
    tags: list[str]
    is_featured: bool
    location: str
    salary_min: int
    salary_max: int
    status: JobStatus
    created_at: datetime
    updated_at: datetime


class JobPublicRead(BaseModel):
    """Schema for public job board responses.

    Excludes internal fields (company_id, status, updated_at) that should
    not be exposed to unauthenticated users. Status is omitted because the
    public endpoint only ever returns PUBLISHED jobs — it carries no
    information and leaks an internal enum.
    """

    model_config = ConfigDict(from_attributes=True)

    id: int
    title: str
    short_description: str
    description: str
    requirements: list[JobRequirementItem]
    tags: list[str]
    is_featured: bool
    location: str
    salary_min: int
    salary_max: int
    created_at: datetime


class JobContactEmailRequest(BaseModel):
    """Schema for admin sending a contextual email to a company about a job."""

    admin_note: str = Field(default="", max_length=2000)

File: frontend/src/types/jobs.ts
import type { JobStatus } from "@/types/enums";

export const JOB_TITLE_MAX = 100;
export const JOB_LOCATION_MAX = 200;
export const JOB_DESC_MAX = 5000;
export const JOB_SHORT_DESC_MAX = 140;
export const JOB_TAG_MAX_LEN = 30;
export const JOB_TAG_MAX_COUNT = 6;
export const JOB_REQ_TEXT_MAX = 200;
export const JOB_REQ_MIN_COUNT = 3;
export const JOB_REQ_MAX_COUNT = 15;

/** Fallback salary bounds used when job list is empty or all salaries are equal. */
export const SALARY_FALLBACK = { min: 0, max: 50_000 } as const;

export interface JobRequirementItem {
  text: string;
}

export interface JobRead {
  id: number;
  company_id: number;
  title: string;
  short_description: string;
  description: string;
  requirements: JobRequirementItem[];
  tags: string[];
  is_featured: boolean;
  location: string;
  salary_min: number;
  salary_max: number;
  status: JobStatus;
  created_at: string;
  updated_at: string;
}

export interface JobCreate {
  title: string;
  short_description: string;
  description: string;
  requirements: JobRequirementItem[];
  tags: string[];
  location: string;
  salary_min: number;
  salary_max: number;
}

export interface JobUpdate {
  title?: string;
  short_description?: string;
  description?: string;
  requirements?: JobRequirementItem[];
  tags?: string[];
  location?: string;
  /** NOT NULL in DB — backend rejects explicit null. Omit to leave unchanged. */
  salary_min?: number;
  /** NOT NULL in DB — backend rejects explicit null. Omit to leave unchanged. */
  salary_max?: number;
  status?: JobStatus;
}

/** Mirrors backend JobAdminCreate schema. */
export interface JobAdminCreate {
  company_id: number;
  title: string;
  short_description: string;
  description: string;
  requirements: JobRequirementItem[];
  tags: string[];
  is_featured?: boolean;
  location: string;
  salary_min: number;
  salary_max: number;
  status?: JobStatus;
}

/** Mirrors backend JobAdminUpdate — extends JobUpdate with is_featured. */
export interface JobAdminUpdate extends JobUpdate {
  is_featured?: boolean;
}

/** Candidate-side info about their own application for the job (Sprint 11 / #606). */
export interface MyApplicationInfo {
  id: number;
  /** True only when the application is still in NEW status — the candidate
   *  can edit / withdraw it. False means admin engaged with it. The raw
   *  status is intentionally hidden from candidate-facing payloads. */
  editable: boolean;
}

/** Mirrors backend JobPublicRead schema. Status is omitted (only PUBLISHED returned).
 *  `my_application` is populated only on the per-job detail endpoint when
 *  the request bears a candidate JWT; otherwise null. */
export interface JobPublicRead {
  id: number;
  title: string;
  short_description: string;
  description: string;
  requirements: JobRequirementItem[];
  tags: string[];
  is_featured: boolean;
  location: string;
  salary_min: number;
  salary_max: number;
  created_at: string;
  my_application: MyApplicationInfo | null;
}
